### DISCUSSION TITLE
Pro Micro Compatibility

### DISCUSSION THREAD

**Discussion by camandbass** [tttapa/Control-Surface#812] [Q&A]
Hello! 

I've been prototyping a midi controller with Control-Surface.h using an arduino uno flashed to be a mocolufa. This prototype has worked well and I love the Control-Surface library. However, when I got in a set of Pro Micro boards to act in the final product, my code is no longer working. I've added MIDIUSB.h and USBMIDI.h but neither have solved my issue. Here is my code:

```cpp
#include <usbmidi.h>
#include <frequencyToNote.h>
#include <MIDIUSB.h>
#include <MIDIUSB_Defs.h>
#include <pitchToFrequency.h>
#include <pitchToNote.h>




#include <Control_Surface.h> // Include the Control Surface library


// Instantiate a MIDI Debug interface at 115200 baud.
USBMIDI_Interface midi;

 
using namespace MIDI_Notes;
 
// Instantiate a CCButton object
CCPotentiometer pot0 {A0, {MIDI_CC::General_Purpose_Controller_1, CHANNEL_1}};
CCPotentiometer pot1 {A1, {MIDI_CC::General_Purpose_Controller_2, CHANNEL_1}};
CCPotentiometer pot2 {A2, {MIDI_CC::General_Purpose_Controller_3, CHANNEL_1}};
CCPotentiometer pot3 {A3, {MIDI_CC::General_Purpose_Controller_4, CHANNEL_1}};
CCButton button0 = {7, {MIDI_CC::General_Purpose_Controller_5, CHANNEL_1}};
CCButton button1 = {5, {MIDI_CC::General_Purpose_Controller_6, CHANNEL_1}};




constexpr pin_t ledPinb0 = 8;
constexpr pin_t ledPinb1 = 9;
constexpr pin_t ledPinp0 = 10;
constexpr analog_t minimumValue = 16;
constexpr analog_t maximumValue = 16383-16;
analog_t mappingFunction(analog_t raw) {
  raw = constrain(raw, minimumValue, maximumValue);
  return map(raw, minimumValue, maximumValue, 0, 127);
}
void setup() { 
  Control_Surface.begin(); // Initialize Control Surface
  pinMode(ledPinb0, OUTPUT);
  pinMode(ledPinb1, OUTPUT);
  pinMode(ledPinp0, OUTPUT);
}
 
void loop() {
  Control_Surface.loop(); // Update the Control Surface
  if (button0.getButtonState() == Button::Rising)
    digitalWrite(ledPinb0, LOW);
  else if(button0.getButtonState() == Button::Falling)
    digitalWrite(ledPinb0, HIGH);
    delay(1);

 if (button1.getButtonState() == Button::Rising)
    digitalWrite(ledPinb1, LOW);
  else if(button1.getButtonState() == Button::Falling)
    digitalWrite(ledPinb1, HIGH);
    delay(1);


  analogWrite(ledPinp0, pot0.getValue());
  delay(1);

   


  
  

}
```

There are 3 regular pots, one slide pot with an led and two clicky buttons with leds. The led on the slide should show its value through brightness and the two buttons should light their leds on press. If anyone is familiar with working on these pro micro boards any insight would be greatly appreciated!!

All the Best,
Chris

**Comment by tttapa** [comment]
>  I've added MIDIUSB.h and USBMIDI.h but neither have solved my issue.

Simply adding the header files won't affect how Control Surface handles MIDI.

The selection of the appropriate MIDI USB backend happens here:
https://github.com/tttapa/Control-Surface/blob/d16a6fbd7dbe18ac8ed037a4aa6033e603b0aa11/src/MIDI_Interfaces/USBMIDI/USBMIDI.hpp#L36-L81

The Pro Micro is not an officially supported board, so you might have to add some extra checks to make sure that the `"USBMIDI_MIDIUSB.hpp"` backend is selected.

> ```cpp
> // Instantiate a MIDI Debug interface at 115200 baud.
> USBMIDI_Interface midi;
> ```

Probably stating the obvious, but the comment is inaccurate.

> ```cpp
> analog_t mappingFunction(analog_t raw) {
>   raw = constrain(raw, minimumValue, maximumValue);
>   return map(raw, minimumValue, maximumValue, 0, 127);
> }
> ```

This function on its own has no effect. If you want to map your potentiometer, see [CCPotentiometer-Map.ino](https://tttapa.github.io/Control-Surface-doc/Doxygen/d3/d26/CCPotentiometer-Map_8ino-example.html).

> ``` cpp
> delay(1);
> ```

Why?

> ```cpp
> analogWrite(ledPinp0, pot0.getValue());
> ```

`pot0.getValue()` returns a 7-bit number, whereas `analogWrite` expects an 8-bit value.

**Reply by camandbass** [reply]
Hey Pieter!

Thanks so much for the timely response!

When it comes to coding I'm basically a monkey slapping at a typewriter so I appreciate your breakdown!

I figured I had to actually use the USBMIDI library in some way but just didn't understand the documentation.

The comment was a leftover from testing the prototype, left in so I could remember how to check the work.

Somehow I was able to use that mapping function to output values that worked with the Uno, but yeah it was probably just doing nothing and the light worked as a fluke or through some process I didn't understand.

Why delay? Habits built on ignorance I imagine...

Somehow between the broken mapping and analogWrite when I turned the pot it was giving values between 0-127 and dimming the light so I thought it was ok.

I've got a lot of reading to do and when/if I get the micro working I'll share my fixed code here and mark this discussion as answered!

All the best,
Chris
